tokio/src/util/intrusive_double_linked_list.rs

```diff
@@ -0,0 +1,783 @@
+    /// Safety: This function is only safe as long as `node` is guaranteed to
+    /// get removed from the list before it gets moved or dropped.
+    /// In addition to this `node` may not be added to another other list before
+    /// it is removed from the current one.
+    pub unsafe fn add_front(&mut self, node: &mut ListNode<T>) {
+        node.next = self.head;
+        node.prev = None;
+        match self.head {
+            Some(mut head) => head.as_mut().prev = Some(node.into()),
+            None => {}
+        };
+        self.head = Some(node.into());
+        if self.tail.is_none() {
+            self.tail = Some(node.into());
+        }
+    }
+
+    /// Inserts a node into the list in a way that the list keeps being sorted.
+    /// Safety: This function is only safe as long as `node` is guaranteed to
+    /// get removed from the list before it gets moved or dropped.
+    /// In addition to this `node` may not be added to another other list before
+    /// it is removed from the current one.
+    pub unsafe fn add_sorted(&mut self, node: &mut ListNode<T>)
+    where
+        T: PartialOrd,
+    {
+        if self.head.is_none() {
+            // First node in the list
+            self.head = Some(node.into());
+            self.tail = Some(node.into());
+            return;
+        }
+
+        let mut prev: Option<NonNull<ListNode<T>>> = None;
+        let mut current = self.head;
+
+        while let Some(mut current_node) = current {
+            if node.data < current_node.as_ref().data {
+                // Need to insert before the current node
+                current_node.as_mut().prev = Some(node.into());
+                match prev {
+                    Some(mut prev) => {
+                        prev.as_mut().next = Some(node.into());
+                    }
+                    None => {
+                        // We are inserting at the beginning of the list
+                        self.head = Some(node.into());
+                    }
+                }
+                node.next = current;
+                node.prev = prev;
+                return;
+            }
+            prev = current;
+            current = current_node.as_ref().next;
+        }
+
+        // We looped through the whole list and the nodes data is bigger or equal
+        // than everything we found up to now.
+        // Insert at the end. Since we checked before that the list isn't empty,
+        // tail always has a value.
+        node.prev = self.tail;
+        node.next = None;
+        self.tail.as_mut().unwrap().as_mut().next = Some(node.into());
+        self.tail = Some(node.into());
+    }
+
+    /// Returns the first node in the linked list without removing it from the list
+    /// The function is only safe as long as valid pointers are stored inside
+    /// the linked list.
+    /// The returned pointer is only guaranteed to be valid as long as the list
+    /// is not mutated
```

Comment:
Is there a reason these cannot mutably borrow the list to ensure this invariant isn't violated?

Reply:
You also can't directly manipulate the list anyway, because that requires a mutable reference, but the peek borrows the list and prevents getting one.

Maybe the comment is misphrased. However what people can do is insert the pointer into another list, which would blow things up. But that is at least an `unsafe` operation, in the same fashion as `remove`.